packages/eslint-plugin/src/util/needsToBeAwaited.ts

```diff
@@ -0,0 +1,44 @@
+import type * as ts from 'typescript';
+
+import {
+  isTypeAnyType,
+  isTypeUnknownType,
+} from '@typescript-eslint/type-utils';
+import * as tsutils from 'ts-api-utils';
+
+export enum Awaitable {
+  Always,
+  Never,
+  May,
+}
+
+export function needsToBeAwaited(
+  checker: ts.TypeChecker,
+  node: ts.Node,
+  type: ts.Type,
+): Awaitable {
+  // `any` and `unknown` types may need to be awaited
+  if (isTypeAnyType(type) || isTypeUnknownType(type)) {
+    return Awaitable.May;
+  }
+
+  // 'thenable' values should always be be awaited
+  if (tsutils.isThenableType(checker, node, type)) {
+    return Awaitable.Always;
+  }
+
+  // recurse into a type parameter constraint
+  if (tsutils.isTypeParameter(type)) {
+    const constraint = type.getConstraint();
+
+    // if there's no constraint, this may need to be awaited
+    if (!constraint) {
+      return Awaitable.May;
+    }
+
+    return needsToBeAwaited(checker, node, constraint);
+  }
+
+  // anything else should not be awaited
+  return Awaitable.Never;
+}
```

Comment:
hmm. We _should_ be able to use `getConstrainedTypeAtLocation` in the caller to solve this, rather than dealing with the type parameter constraint here, but unfortunately that function... doesn't really work on unconstrained generics, since it doesn't return an unknown type. Filed https://github.com/microsoft/TypeScript/issues/60475.

That said, I think let's try to match the implementation here as closely as possible to `getConstrainedTypeAtLocation()`, though... I'm thinking like this?

```suggestion
export function needsToBeAwaited(
  checker: ts.TypeChecker,
  node: ts.Node,
  type: ts.Type,
): Awaitable {
  // can't use `getConstrainedTypeAtLocation` directly since it's bugged for
  // unconstrained generics.
  const constrainedType = !tsutils.isTypeParameter(type)
    ? type
    : checker.getBaseConstraintOfType(type);

  // unconstrained generic types should be treated as unknown
  if (constrainedType == null) {
    return Awaitable.May;
  }

  // `any` and `unknown` types may need to be awaited
  if (isTypeAnyType(constrainedType) || isTypeUnknownType(constrainedType)) {
    return Awaitable.May;
  }

  // 'thenable' values should always be be awaited
  if (tsutils.isThenableType(checker, node, constrainedType)) {
    return Awaitable.Always;
  }

  // anything else should not be awaited
  return Awaitable.Never;
}
```

Note that writing it without recursion (like `getConstrainedTypeAtLocation`) assumes that descending into a constraint is idempotent.... and, if it's not, then we better fix that for `getConstrainedTypeAtLocation` too! (I'm wondering if that's indeed the case for generics whose constraint is another generic 🤔 maybe a bugged test case can be devised for this and we can file a new issue to fix it in both places if so)

Reply:
Should I add a TODO for refactor this to use `getConstrainedTypeAtLocation()` once the `unknown` type is available on `ts.TypeChecker`?

Reply:
I'd say no TODO comment, but just file a repo maintenance followup issue that's a catch-all for investigating edge cases around unconstrained generics in both `getConstrainedTypeAtLocation` and mention this discussion/`needsToBeAwaited`. There's a bit of bug-hunting I want to do with the current impl of `getConstrainedTypeAtLocation` too. 

Unfortunately even if TS adds `checker.getUnknownType()` we still have to wait until we drop support for the latest TS version that doesn't have it yet in order to actually use it, which is >= 2 years from now 🫤